File: Golang_src/easy/majorityElement_169.go
package easy

// Use Boyer-Moore Voting Algorithm:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 16 ms, faster than 96.03%
// Memory Usage: 6 MB, less than 100.00%
func majorityElement(nums []int) int {
	result := nums[0]
	count := 0
	threshold := len(nums)/2 + 1

	for _, v := range nums {
		if result == v {
			count++
		} else {
			count--
		}
		if count < 0 {
			result = v
			count = 1
		}
		if count == threshold {
			return result
		}
	}

	return result
}

File: Cpp_src/easy/majority_element_169.cc
#include <iostream>
#include <vector>
using namespace std;

// Use Boyer-Moore vote:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 16 ms, faster than 96.51%
// Memory Usage: 8.8 MB, less than 100.00%
class Solution
{
public:
    int majorityElement(vector<int> &nums)
    {
        int result = nums[0], cnt = 0;
        for (int num : nums)
        {
            if (num == result)
            {
                cnt++;
                if (cnt > nums.size())
                {
                    return result;
                }
            }
            else
            {
                cnt--;
                if (cnt == 0)
                {
                    result = num;
                    cnt = 1;
                }
            }
        }

        return result;
    }
};